rust/parquet/src/data_type.rs

```diff
@@ -392,27 +414,109 @@ macro_rules! make_type {
             fn get_type_size() -> usize {
                 $size
             }
+
+            fn get_column_reader(
+                column_writer: ColumnReader,
+            ) -> Option<ColumnReaderImpl<Self>> {
+                match column_writer {
+                    ColumnReader::$reader_ident(w) => Some(w),
+                    _ => None,
+                }
+            }
+
+            fn get_column_writer(
+                column_writer: ColumnWriter,
+            ) -> Option<ColumnWriterImpl<Self>> {
+                match column_writer {
+                    ColumnWriter::$writer_ident(w) => Some(w),
+                    _ => None,
+                }
+            }
+
+            fn get_column_writer_ref(
+                column_writer: &ColumnWriter,
+            ) -> Option<&ColumnWriterImpl<Self>> {
+                match column_writer {
+                    ColumnWriter::$writer_ident(w) => Some(w),
+                    _ => None,
+                }
+            }
+
+            fn get_column_writer_mut(
+                column_writer: &mut ColumnWriter,
+            ) -> Option<&mut ColumnWriterImpl<Self>> {
+                match column_writer {
+                    ColumnWriter::$writer_ident(w) => Some(w),
+                    _ => None,
+                }
+            }
         }
     };
 }
 
 // Generate struct definitions for all physical types
 
-make_type!(BoolType, Type::BOOLEAN, bool, 1);
-make_type!(Int32Type, Type::INT32, i32, 4);
-make_type!(Int64Type, Type::INT64, i64, 8);
-make_type!(Int96Type, Type::INT96, Int96, mem::size_of::<Int96>());
-make_type!(FloatType, Type::FLOAT, f32, 4);
-make_type!(DoubleType, Type::DOUBLE, f64, 8);
+make_type!(
+    BoolType,
+    Type::BOOLEAN,
+    BoolColumnReader,
```

Comment:
I am not sure about this. We generate column readers and writers when making types?

Reply:
It generates the unpacking functions needed to unpack the enum. I don't know of another, better way of doing this so 🤷‍♂

Reply:
Well, this is just a weird place to put them.

Reply:
https://github.com/Marwes/arrow/commit/eb1ba90e3434bfd11b9f9d5be3c55d90d377e81f is the only alternative I know which I would argue is a lot more verbose for no gain.